### ISSUE TITLE
AptX Adaptive support #2

### ISSUE THREAD

**Issue by leleobhz** [Arkq/openaptx#8]
Following up #1 to track aptX Adaptive support.

**Comment by leleobhz** [comment]
@Arkq do you already know the .so and binary files that do aptX Adaptive in a phone?

**Comment by Arkq** [comment]
The only library that I've got is this one: https://github.com/Arkq/openaptx/blob/master/archive/qdsp6/capi_v2_aptX_CLHDAD_Encoder.so Unfortunately, it's for QDSP6 platform which is rater nor suitable (for me) for reverse engineering.

**Comment by anonymix007** [comment]
@Arkq you'll probably won't find ARM or X86 versions of aptX Adaptive library. However, QDSP6 (aka Hexagon DSP) is kinda supported in [Ghidra](https://github.com/toshipiazza/ghidra-plugin-hexagon) and in [Binary Ninja](https://github.com/google/binja-hexagon).
[Here](https://github.com/Arkq/openaptx/files/9376469/QDSP6.zip) are newer V2.2 and V3 QDSP6 encoder libraries. Looks like   V2.2 is the latest aptX Adaptive version, and V3 is called aptX Lossless.
As for decoder libraries, it's worse. I do have object files for their Kalimba architecture 4 (used in QCC Bluetooth SoCs), but some of them may be stripped. There exists disassembler for this architecture (well, actually full GNU binutils build is also available as well as compiler).

**Comment by Arkq** [comment]
The problem with QDSP6 is that the algorithm in there is not as straightforward as in x86/ARM code. I've been checking it in IDA but I was not able to get anything useful (in a timeframe I can invest in that).

**Comment by anonymix007** [comment]
Can you also take a look at [Kalimba](https://github.com/Arkq/openaptx/files/9377137/kalimba.zip) versions? I added few object files as well as objdump binary.
Maybe it will be easier to add support of Kalimba to i.e. Ghidra and only after that try to reverse it.

**Comment by Arkq** [comment]
> Can you also take a look at Kalimba versions?

Yes, these binaries might be reversible at least based on symbols in them. Anyway Kalimba is not a general purpose CPU, so the assembler code might not be easy to read.

**Comment by anonymix007** [comment]
@arkq I think I found x86 executables for decoding aptX Adaptive (looks like this includes aptX Lossless as well). Would you be interested in taking a look into them?

**Comment by arkq** [comment]
Yes, I can take a look at it. You can send it on my email or submit PR for this repo with lib added to archives directory. I'm not sure whether I will be able to reverse it to C code in a reasonable time period (I've got lots of other things on my head currently), but at leas I will check how much effort will be required :)

**Comment by anonymix007** [comment]
@arkq PR was submitted. It will probably be a lot of effort, maybe even more than kalimba ones (they at least have the debug information)

**Comment by Monke132** [comment]
Do we have progress on this?

**Comment by anonymix007** [comment]
Nope. Maybe if there was Ghidra for Kalimba, it wouldn't be hard to reverse engineer. 
Using those executables directly (inside of the emulator) might be easier until someone will add it to Ghidra or IDA (ISA docs are not public tho I have them).

**Comment by DrMirage** [comment]
is there any new progress?

**Comment by Animatron1** [comment]
The silence on this topic is quite deafening... It's been years...

**Comment by moracabanas** [comment]
Yeah, I know Steam Deck added aptX LL support but, Bose eaburds didn't have aptX support until Bose Quiet Confort (Ultra) skipping completely LL. 
Because they aren't backward compatible with LL, most Adaptivr earbuds render useless to play games, and It is just another lost battle to the Bluetooth monopoly...
I hope reverse engineering wins this battle!

**Comment by Rongronggg9** [comment]
Some AOSP vendor components \[1] reveal how to parse the Codec Specific Information Elements of AptX Adaptive. Some other AOSP vendor components contain the configuration for the HW DSP encoder \[2]\[3].

(Quickly glancing at these codes, I doubt if the actual AptX Adaptive encoder reuses AptX HD/LL encoders with some modifications to make them "adaptive"...)

\[1]: https://github.com/syberia-project/platform_hardware_qcom_bt/blob/439c224484e233c4ff189b19d47a16a65fe1fc5a/bthost_ipc/bthost_ipc.c#L549

\[2]: https://github.com/sonyxperiadev/vendor-qcom-opensource-interfaces/blob/764d41a3fcdf490df2eb4a9c347184e02e2895c7/bluetooth_audio/2.0/types.hal#L164

\[3]: https://git.codelinaro.org/clo/le/platform/hardware/qcom/audio/-/blob/LA.UM.12.2.1.c26-00500-sdm660.0/hal/audio_extn/a2dp.c?ref_type=tags

**Comment by Tseshongfeeshur** [comment]
@arkq [These files](https://github.com/user-attachments/files/25719944/aptx-adaptive-patch.tar.gz) were extracted from a Xiaomi 12X, which officially supports aptX Adaptive.
I suspect the aptX Adaptive implementation is contained in them somewhere, but my skills are quite limited and I couldn’t confirm it properly.
Would you mind taking a look? I’d really appreciate it! 🌹

**Comment by anonymix007** [comment]
> I suspect the aptX Adaptive implementation is contained in them somewhere

Funnily enough, it is not. aptX Adaptive on these devices is running on the ADSP, and the libraries are usually stored on a separate partition (`/dev/block/by-name/dsp_*`). These are built for QDSP6 and reverse engineering them would be really painful. I doubt anyone would work on that for free.

Anyway, I think the best option here would be to use those libraries as-is and run them in QEMU or  maybe even natively on Snapdragon X1/X2 laptops.

**Comment by moracabanas** [comment]
I would just switch my free time ML developer carreer to reverse engineer this codec just to be able to use APTX Adaptive on the Steam Deck.

Looking for a second hand Snapdragon X laptop to start with 😂

**Comment by Tseshongfeeshur** [comment]
Sorry for my useless advice...
In fact, I'm trying to figure out how to restore aptX Adaptive support from the stock HyperOS to crDroid, and of course, I've failed so far. 😭

**Comment by anonymix007** [comment]
> aptX Adaptive support from the stock HyperOS to crDroid

That's actually the easy part. You just don't use LineageOS/AOSP based ROMs but rather go for CLO based ones. If you only care for Bluetooth then it should be enough to just build [CLO BT stack](https://git.codelinaro.org/clo/la/platform/packages/modules/Bluetooth/-/tree/ks-aosp.lnx.16.0.r1-rel?ref_type=heads). AFAIR since A15 it's just an APEX anyway, so replacing that repo in your AOSP ROM tree, fixing all build errors and rebuilding should be enough. For the reference, it's basically how my https://github.com/anonymix007/ExtA2DP module (which I hope to come back to one day but probably won't) worked. But in your case I'd strongly recommend you to rebuild the ROM itself.

> I would just switch my free time ML developer carreer to reverse engineer this codec just to be able to use APTX Adaptive on the Steam Deck.
> 
> Looking for a second hand Snapdragon X laptop to start with 😂

I can only wish you good luck. QEMU route should not require any reverse engineering, some C knowledge would be enough. Snapdragon X it has user-programmable CDSP. What one needs to develop is a some sort of [AudioReach module](https://audioreach.github.io) loader for CDSP. I assume it should be possible to run almost the same code in QEMU as well, but it'll probably take some effort to set up. 
I once wanted to work on this by myself but lack of time, work and the fact that I really had no use for this didn't help.

**Comment by arkq** [comment]
> These are built for QDSP6 and reverse engineering them would be really painful. I doubt anyone would work on that for free.

That's exactly the point here. Once I've looked at these libraries, but the code is "optimized" for QDSP6 which obfuscates the logic behind it. Even though I known what to expect more or less, I'm not familiar enough with QDSP6 asembler code to see what is going on in there :D

**Comment by AndyCruz08** [comment]
[aptxlibs_CPH2749.zip](https://github.com/user-attachments/files/29668504/aptxlibs_CPH2749.zip) some more aptX adaptive blobs from an OTA update for a phone, apparently there is a version 4 of aptX adaptive, not sure what that's about.  still QDSP6/Hexagon, this time its version 79 which isn't supported by Qemu (and might not be fully supported by Ghidra, still opens fine for me) but does have reference manuals available, hope this still helps! Includes libraries related to decoding aptX Adaptive 4 (likely for voice only, at least that's what I guessed from lack of references to other modes in Ghidra) and possible helper libraries (prefixed with spf) for encoding v3 for ble and v4 for xpan, and decoding v4 for xpan

**Comment by littletakbest** [comment]
> [aptxlibs_CPH2749.zip](https://github.com/user-attachments/files/29668504/aptxlibs_CPH2749.zip) some more aptX adaptive blobs from an OTA update for a phone, apparently there is a version 4 of aptX adaptive, not sure what that's about. still QDSP6/Hexagon, this time its version 79 which isn't supported by Qemu (and might not be fully supported by Ghidra, still opens fine for me) but does have reference manuals available, hope this still helps! Includes libraries related to decoding aptX Adaptive 4 (likely for voice only, at least that's what I guessed from lack of references to other modes in Ghidra) and possible helper libraries (prefixed with spf) for encoding v3 for ble and v4 for xpan, and decoding v4 for xpan

how did you extract those? since I own a xiaomi 15 so I can get you the binaries if needed

**Comment by AndyCruz08** [comment]
its been a while, but iirc you have to extract something like `dsp.img` or `adsp.img` from the firmware files (i used a oneplus ota for this), has to be a phone that supports aptX adaptive out of the box ofc

**Comment by baizhu945** [comment]
I just implemented aptx adaptive using pipewire on my Linux, please check out my latest PR https://github.com/arkq/openaptx/pull/12 . But I haven't connect to a headphone which supports aptx adaptive yet , and this need deeper test .
